bootstrap/src/test.rs

```diff
@@ -0,0 +1,157 @@
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+use anstream::{eprint as print, eprintln as println};
+use clap::Args;
+use color_print::{cprint, cprintln};
+use glob::glob;
+use which::which;
+
+use crate::manifest::Manifest;
+
+/// Run tests
+#[derive(Args, Debug)]
+pub struct TestCommand {}
+
+impl TestCommand {
+    pub fn run(&self, manifest: &Manifest) {
+        manifest.prepare();
+
+        std::panic::set_hook(Box::new(|info| {
+            cprintln!("<r,s>Test failed</r,s>: {}", info);
+        }));
+
+        let testcases = self.collect_testcases(manifest);
+        cprintln!("<b>[TEST]</b> found {} testcases", testcases.len());
+
+        let filechecker = FileChecker::new();
+        for testcase in testcases {
+            match testcase.test {
+                TestType::FileCheck => {
+                    cprint!("File checking {}...", testcase.name);
+                    testcase.build(manifest);
+                    filechecker.run(&testcase.source, &testcase.output);
+                }
+                TestType::Compile => {
+                    cprint!("Compiling {}...", testcase.name);
+                    testcase.build(manifest);
+                }
+            }
+            cprintln!("<g>OK</g>");
+        }
+    }
+
+    pub fn collect_testcases(&self, manifest: &Manifest) -> Vec<TestCase> {
+        let mut result = vec![];
+
+        // Examples
+        for case in glob("example/*.rs").unwrap() {
+            let case = case.unwrap();
+            let filename = case.file_stem().unwrap();
+            if filename == "mini_core" {
+                continue;
+            }
+            let name = format!("example/{}", filename.to_string_lossy());
+            let output = manifest.out_dir.join("example").join(Path::new(filename));
```

Comment:
Suggestion [JOIN-PATH 1/2]: I want to say `join` takes a `AsRef<Path>` so I think you don't need `Path::new` step.

Reply:
Fixed.